src/node_api.cc

```diff
@@ -492,15 +477,41 @@ class TryCatch : public v8::TryCatch {
 
 //=== Function napi_callback wrapper =================================
 
-static const int kDataIndex = 0;
-static const int kEnvIndex = 1;
+static const int kFunctionIndex = 0;  // Used in CallbackBundle::cb[]
+static const int kGetterIndex = 0;    // Used in CallbackBundle::cb[]
+static const int kSetterIndex = 1;    // Used in CallbackBundle::cb[]
+static const int kCallbackCount = 2;  // Used in CallbackBundle::cb[]
+                                      // Max is "getter + setter" case
+
+static const int kCallbackBundleIndex = 0;  // The first and the only one
+static const int kInternalFieldCount = 1;
 
-static const int kFunctionIndex = 2;
-static const int kFunctionFieldCount = 3;
+// Use this data structure to reduce the number
+// of GetInternalField() calls to only 1 (was: 3).
+// This leads to better performance in runtime.
+// Ref: benchmark/misc/napi_function_call
+struct CallbackBundle {
+  // Bind the lifecycle of `this` C++ object to a JavaScript object.
+  // We never delete a CallbackBundle C++ object directly.
+  void BindLifecycleTo(v8::Isolate* isolate, v8::Local<v8::Value> target) {
+    handle.Reset(isolate, target);
+    handle.SetWeak(this, WeakCallback, v8::WeakCallbackType::kParameter);
+  }
+
+  napi_env       env;      // Necessary to invoke C++ NAPI callback
+  void*          cb_data;  // The user provided callback data
+  napi_callback  cb[kCallbackCount];   // Max capacity is 2 (getter + setter)
+  node::Persistent<v8::Value> handle;  // Die with this JavaScript object
 
-static const int kGetterIndex = 2;
-static const int kSetterIndex = 3;
-static const int kAccessorFieldCount = 4;
+ private:
+  static void WeakCallback(v8::WeakCallbackInfo<CallbackBundle> const& info) {
+    // Use WeakCallback mechanism to delete the C++ `bundle` object.
+    // This will be called when object in `handle` is being GC-ed.
+    if (CallbackBundle* bundle = info.GetParameter()) {
```

Comment:
This needs to be split up:
```C++
CallbackBundle* bundle = info.GetParameter();
if (bundle != nullptr) {
...
```

Reply:
My bad, that was my ancient habit to declare a pointer in `if`.

Actually we can directly delete the pointer without checking whether it's null (deleting a null pointer is explicitly defined to have no effect in 14882)

Reply:
👍 Better still.